src/include/serial_inline.h

```diff
@@ -300,32 +299,12 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_PAGE *page
         return (0);
 
     /*
-     * We would like to call __wt_txn_update_oldest only in the event that there are further updates
-     * to this page, the check against WT_TXN_NONE is used as an indicator of there being further
-     * updates on this page.
+     * Look for obsolete updates if the page size is more than 50% of the maximum split page size
+     * allowed by btree. This is to prevent unnecessary page splits due to the obsolete updates
+     * present on the page.
      */
-    if ((txn = page->modify->obsolete_check_txn) != WT_TXN_NONE) {
-        obsolete_timestamp = page->modify->obsolete_check_timestamp;
-        if (!__wt_txn_visible_all(session, txn, obsolete_timestamp)) {
-            /* Try to move the oldest ID forward and re-check. */
-            ret = __wt_txn_update_oldest(session, 0);
-            /*
-             * We cannot proceed if we fail here as we have inserted the updates to the update
-             * chain. Panic instead. Currently, we don't ever return any error from
-             * __wt_txn_visible_all. We can catch it if we start to do so in the future and properly
-             * handle it.
-             */
-            if (ret != 0)
-                WT_RET_PANIC(session, ret, "fail to update oldest after serializing the updates");
-
-            if (!__wt_txn_visible_all(session, txn, obsolete_timestamp))
-                return (0);
-        }
-
-        page->modify->obsolete_check_txn = WT_TXN_NONE;
-    }
-
-    __wt_update_obsolete_check(session, cbt, upd->next, true);
+    if (page->memory_footprint > (S2BT(session)->splitmempage / 2))
```

Comment:
I can imagine cases where there are lots of small updates accumulating that could otherwise be released, while rarely triggering this 50% threshold.

You've mentioned before that we don't keep walking past necessary updates, so maybe that isn't a performance concern, but it will alter how much space such a page consumes in cache (meaning there is less space for other things). Would you be amenable to using a combination of your suggested approach and the one from @keitharnoldsmith of doing a random sampling?

Reply:
Yes, I agree with you that the small updates on a page can be accumulated for a longer time. But these pages are being cleaned for obsolete updates as part of the reconciliation, even if their size has not reached 50% of the page size.

We can do both the random or 50% off size as a condition to enter an obsolete check.